Draft of 302 pytorch nncf quantization

ljaljushkin opened the pull request:
The goal of this notebook to demonstrate how to use NNCF 8-bit quantization to optimize the PyTorch model for inference with OpenVINO Toolkit.

ryanloney commented:
Putting my notes here so @helena-intel can review when she is back. A few open items: 

- If we put "!pip install" inside the notebook (for Colab usage), we should use openvino-dev[onnx] so all cells can execute
- On Windows 10, the notebook fails at the import step with the following: 
`CalledProcessError: Command '['where', 'cl']' returned non-zero exit status 1.`
(this seems to be an issue with how subprocess is used) 
- When running inside the UBI 8 docker container, I needed to run docker with `--shm-size 8G` otherwise the training step fails
- Alexander suggests reducing `batch_size` parameter to 128 to avoid the docker shared memory issue above. He said it does not impact the performance and training/tuning time and accuracy results should be close 


This is a summary of results when I ran the notebook on various systems: 

**Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz:**
- **OS:** macOS 11.4
- **Training time:** 30min
- **FP32:** 770 FPS
- **INT8:** 1961 FPS

**Intel(R) Core(TM) i7-10700K CPU @ 3.80GHz:**
- **OS:** Red Hat UBI 8 (container)
- **Training time:** 8min
- **FP32:** 1277 FPS
- **INT8:** 5493 FPS

**Intel(R) Core(TM) i7-10700K CPU @ 3.80GHz:**
- **OS:** Ubuntu 18.04 LTS 
- **Training time:** 8min
- **FP32:** 1301 FPS
- **INT8:** 5061 FPS

After re-running the Ubuntu benchmark, the results are similar to UBI 8 (~5400 FPS for INT8). But for some reason the first result was not as good. 

On all systems, the accuracy is slightly better for the INT8 model after fine tuning.

ryanloney reviewed (changes requested):
Please review and accept suggestions. Most suggestions are just for readability, but if any of my suggestions incorrectly changes meaning, please reject or add a comment to clarify.

helena-intel reviewed (changes requested):
Thanks for this contribution @ljaljushkin , it is very welcome to have an NNCF demo. It works for me on Windows too - after installing VS Build Tools and adding some folders to the path. I can document that and make it work in the CI.

Two small requests:
- Can you remove the installation part and add nncf[torch] to our main requirements.txt? 
- Can you move the imports to the top?

I saw that you added the resnet weights to the repo. 43MB is large. Can we store these weights somewhere else?

It would be helpful if you could add a bit more context at the top of the notebook about what this notebook does - that we start with a pretrained model and then finetune with one NNCF. You could also link to the NNCF paper.

ljaljushkin commented:
> more context at the top of the notebook about what this notebook does



> Thanks for this contribution @ljaljushkin , it is very welcome to have an NNCF demo. It works for me on Windows too - after installing VS Build Tools and adding some folders to the path. I can document that and make it work in the CI.
> 
> Two small requests:
> 
> * Can you remove the installation part and add nncf[torch] to our main requirements.txt?
> * Can you move the imports to the top?
> 
> I saw that you added the resnet weights to the repo. 43MB is large. Can we store these weights somewhere else?
> 
> It would be helpful if you could add a bit more context at the top of the notebook about what this notebook does - that we start with a pretrained model and then finetune with one NNCF. You could also link to the NNCF paper.

Addressed requests.
Also added resnet weights to [Git Large File Storage](https://git-lfs.github.com/) and added more context about what this notebook does.

ljaljushkin commented:
@helena-intel Ci fails with strange error `NameError: name 'fp32_pth_path' is not defined` that I can't reproduce locally.
It's strange because this variable is defined in the first cell with imports. Could you please take a look at the issue?

ljaljushkin commented on notebooks/302-pytorch-quantization-aware-training/302-pytorch-quantization-aware-training.ipynb:
```diff
@@ -0,0 +1,818 @@
+    "* Measure and compare the performance of models\n",
+    "For more advanced usage, please refer to these [examples](https://github.com/openvinotoolkit/nncf/tree/develop/examples).\n",
+    "\n",
+    "We selected the ResNet-18 model with Tiny ImageNet-200 dataset. ResNet-18 is the version of ResNet models that contains the fewest layers (18) and Tiny ImageNet-200 is a subset of the larger ImageNet dataset. Using the smaller model and dataset will speed up training and download time. To see other ResNet models, visit [PyTorch hub](https://pytorch.org/hub/pytorch_vision_resnet/)."
```
@helena-intel, would it be enough to satisfy legal requirements?
```suggestion
    "We selected the ResNet-18 model with Tiny ImageNet-200 dataset. ResNet-18 is the version of ResNet models that contains the fewest layers (18) and Tiny ImageNet-200 is a subset of the larger ImageNet dataset that should be downloaded in the notebook. Using the smaller model and dataset will speed up training and download time. To see other ResNet models, visit [PyTorch hub](https://pytorch.org/hub/pytorch_vision_resnet/)."
```

ljaljushkin commented on notebooks/302-pytorch-quantization-aware-training/302-pytorch-quantization-aware-training.ipynb:
```diff
@@ -0,0 +1,821 @@
+    "int8_onnx_path = Path(OUTPUT_DIR / (BASE_MODEL_NAME + '_int8')).with_suffix('.onnx') \n",
+    "int8_ir_path = int8_onnx_path.with_suffix(\".xml\")\n",
+    "\n",
+    "fp32_pth_url = 'https://storage.openvinotoolkit.org/repositories/nncf/models/openvino_notebooks/302_resnet18_fp32.pth'\n",
```
@helena-intel, the notebook doesn't work until this link becomes valid

helena-intel reviewed (approved):
Thanks very much for contributing this notebook @ljaljushkin ! I will merge it now and add a note about Windows. We will add instructions for Windows soon.

ryanloney commented:
Thank you @ljaljushkin and @helena-intel! This is a very exciting notebook, really happy we could merge it and use storage.openvinotoolkit.org. And thanks Helena for finding ways to get it working on Windows (soon).
